algebra/group.lean

```diff
@@ -338,6 +338,17 @@ instance [add_monoid α] : monoid (multiplicative α) :=
   mul_one := @add_zero _ _,
   ..multiplicative.semigroup }
 
+instance {α} : monoid (list α) :=
+{ one := [],
+  mul := λ x y, x ++ y,
+  mul_one := by intros; apply list.append_nil,
+  one_mul := by intros; refl,
+  mul_assoc := by intros; apply list.append_assoc }
+
+@[simp] lemma list.one_def {α} : (1 : list α) = [] := rfl
+
+@[simp] lemma list.mul_def {α} (xs ys : list α) : xs * ys = xs ++ ys := rfl
```

Comment:
I don't think we want a global instance of monoid for list. Maybe you could argue it's an add_monoid, but it's definitely not a multiplicative monoid. I think this was one of the justifying instances of notation-free monoids, because using 1 and * for lists is weird.

Reply:
Yes, I think the `*` and `+` is more like the lifted version, where the `*` from the elements is lifted: `xs * ys = (*) <$> xs <*> ys`

Reply:
What do you guys think of using a type synonym for this? Something like `free_monoid` or `free_add_monoid`. Alternatively, it can also work as a local instance.

Reply:
I'm okay with adding `free_(add_)monoid`, but only for type class instances. I don't think we should add any functions / constants on it. The `free_add_monoid` is a canonically ordered monoid.

Reply:
Do you use the order on lengths to define the monoid's order?

Reply:
You use divisibility. Any monoid is canonically ordered in exactly one way (that's why it's called canonical), and the axiom says what it is. In the case of lists it is right extension, i.e. `xs <= xs ++ ys`.

Reply:
Why extension on the right? Isn't the fact that extension on the left would also yield a proper partial order undermine the idea that the order would be unique?

Reply:
unique in satisfying the axiom of a `canonically_ordered_monoid`, which is right biased

Reply:
and it's called canonical not natural :-)

Reply:
I think I must be out of my depth here. What is the distinction between natural and canonical?

Reply:
Actually, I don't know if there is a distinction...